File: clients/fsharp/Model/PlayerView.fs
#nowarn "0058"

namespace Komus24
namespace Komus24.Model

open Komus24

/// Current game's state
type PlayerView = {
    /// Current tick number
    CurrentTick: int;
    /// Your player
    You: Model.Player;
    /// Other players
    Other: Model.Player[];
    /// Available quests
    Quests: Model.Quest[];
} with

    /// Write PlayerView to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write this.CurrentTick
        this.You.writeTo writer
        writer.Write this.Other.Length
        this.Other |> Array.iter (fun value ->
            value.writeTo writer )
        writer.Write this.Quests.Length
        this.Quests |> Array.iter (fun value ->
            value.writeTo writer )
        ()

    /// Read PlayerView from reader
    static member readFrom(reader: System.IO.BinaryReader) = {
        CurrentTick = reader.ReadInt32()
        You = Model.Player.readFrom reader;
        Other = [|for _ in 1 .. reader.ReadInt32() do
                    yield Model.Player.readFrom reader; |]
        Quests = [|for _ in 1 .. reader.ReadInt32() do
                     yield Model.Quest.readFrom reader; |]
    }

File: clients/rust/src/model/player_view.rs
use super::*;

/// Current game's state
#[derive(Clone, Debug)]
pub struct PlayerView {
    /// Current tick number
    pub current_tick: i32,
    /// Your player
    pub you: model::Player,
    /// Other players
    pub other: Vec<model::Player>,
    /// Available quests
    pub quests: Vec<model::Quest>,
}

impl trans::Trans for PlayerView {
    fn write_to(&self, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        self.current_tick.write_to(writer)?;
        self.you.write_to(writer)?;
        self.other.write_to(writer)?;
        self.quests.write_to(writer)?;
        Ok(())
    }
    fn read_from(reader: &mut dyn std::io::Read) -> std::io::Result<Self> {
        let current_tick: i32 = trans::Trans::read_from(reader)?;
        let you: model::Player = trans::Trans::read_from(reader)?;
        let other: Vec<model::Player> = trans::Trans::read_from(reader)?;
        let quests: Vec<model::Quest> = trans::Trans::read_from(reader)?;
        Ok(Self {
            current_tick,
            you,
            other,
            quests,
        })
    }
}